Fix issue #1712

JohanEngelen opened the pull request:
Cross-module inlining (i.e. emitting some code from another module) is only done when a function is not fully semantic3 analysed.
The problem is that sometimes, a function has already been fully "semantic3'ed", but is not emitted anywhere. This happens for nested functions, and those are currently not cross-module inlined (see #1603).
In issue #1712, it happens for a function with template parameter `__FILE__`, i.e. a function that should always be inlined (#1703) exactly because of the link problem observed in #1712.

This fix for #1712 attempts to solve this for now, by emitting `pragma(inline, true)` functions more aggressively. It fixes the current issue, and I hope it does not create new bugs.

Currently dustmiting to create a testcase; it will take a lot of time.

kinke commented:
Any progress? Otherwise I'd suggest disabling cross-module inlining by default (or only enabling it for optimization levels > 3) for now, until the issue is fully understood.

JohanEngelen commented:
Found the problem (it's bad), fixing it.

JohanEngelen commented:
The problem was that inside `DtoDefineFunction`, a call to `DtoResolveFunction` would end up invoking `DtoDefineFunction` and defining the function with `available_externally`, which would not be "fixed up" later because it all happened in the call trace from the `DtoDefineFunction` that should do the fix up. The fix is to tell `DtoResolveFunction` and `DtoDeclareFunction` that they may _not_ end up defining the function.

JohanEngelen commented:
(I'm still dustmiting the testcase btw. It takes forever)

JohanEngelen commented:
Dustmiting is done, and a fairly simple testcase resulted. I've modified it such that it doesn't depend on Phobos and added it to the PR.
Further fixes were needed involving with nested functions and nested structs. The fix for this issue uncovers unimplemented nested function cross-module emitting problems.

kinke commented:
With every new fix, a new issue pops up. ;/
I'd really recommend not enabling cross-module inlining by default for v1.1 yet, and exposing it via optimization levels > 3 for now.

JohanEngelen commented:
Yeah, I think you're right :(

JohanEngelen commented on gen/function-inlining.cpp:
```diff
@@ -143,18 +143,21 @@ bool defineAsExternallyAvailable(FuncDeclaration &fdecl) {
+  if ((fdecl.inlining != PINLINEalways) && alreadyOrWillBeDefined(fdecl)) {
     // This check is needed because of ICEs happening because of unclear issues
     // upon changing the codegen order without this check.
+    // pragma(inline, true) functions remain defineAsExternallyAvailable
```
The problem is non-singleobj compilation, because most (all?) instantiated functions are put into one module, which may be different from the callsite's module. So if alreadyOrWillBeDefined  is true the function will be defined, yes, but for inlining it needs to be defined in every calling module.

JohanEngelen commented on gen/declarations.cpp:
```diff
@@ -206,10 +206,15 @@ class CodegenVisitor : public Visitor {
+    // If we reach here during codegen of an available_externally function,
+    // the struct initializer should stay external and therefore must not
+    // have an initializer.
+    if ((decl->getModule() == gIR->dmodule) || decl->isInstantiated()) {
```
Definitely.

JohanEngelen commented:
I opened #1737 to "fix" all cross-module inlining issues: disable the thing.

I'd like to enable it at `-O4`, but it's up for debate. I think it's nice to be able to compile with -O4 such that compilation also works with earlier LDC versions (compared with the previously unknown `-enable-cross-module-inlining`) and get the boost of cross-module inlining when compiler with the new LDC. (e.g. in our own CMake file, I imagine using `-O4` when LDC is detected)

Upon 1.1.0 release, I'd like to enable cross-module inlining per default again in master, so it gets more testing exposure.
